cmd/rpcdaemon/commands/trace_api_filtering.go

```diff
@@ -0,0 +1,387 @@
+	// TODO(tjayrush): Parity intersperses block/uncle reward traces with transaction call traces. We need to be able to tell the
+	// difference. I do that with this boolean array. This will be re-written shortly. For now, we use this simple boolean flag.
+	// Eventually, we will fill the ParityTrace array directly as we go (and we can probably even do better than that)
+	var traceTypes []bool
+	var maxTracesCount uint64
+	var offset uint64
+	var skipped uint64
+
+	sort.Slice(req.FromAddress, func(i int, j int) bool {
+		return bytes.Compare(req.FromAddress[i].Bytes(), req.FromAddress[j].Bytes()) == -1
+	})
+
+	sort.Slice(req.ToAddress, func(i int, j int) bool {
+		return bytes.Compare(req.ToAddress[i].Bytes(), req.ToAddress[j].Bytes()) == -1
+	})
+
+	var fromBlock uint64
+	var toBlock uint64
+	if req.FromBlock == nil {
+		fromBlock = 0
+	} else {
+		fromBlock = uint64(*req.FromBlock)
+	}
+
+	if req.ToBlock == nil {
+		headNumber := rawdb.ReadHeaderNumber(api.dbReader, rawdb.ReadHeadHeaderHash(api.dbReader))
+		toBlock = *headNumber
+	} else {
+		toBlock = uint64(*req.ToBlock)
+	}
+
+	if fromBlock > toBlock {
+		// TODO(tjayrush): Parity reports no error in this case it simply returns an empty response
+		return nil, nil //fmt.Errorf("invalid parameters: toBlock must be greater than fromBlock")
+	}
+
+	if req.Count == nil {
+		maxTracesCount = api.maxTraces
+	} else {
+		maxTracesCount = *req.Count
+	}
+
+	if req.After == nil {
+		offset = 0
+	} else {
+		offset = *req.After
+	}
+
+	if err := api.db.View(ctx, func(tx ethdb.Tx) error {
+		if req.FromAddress != nil || req.ToAddress != nil { // use address history index to retrieve matching transactions
+			var historyFilter []*common.Address
+			isFromAddress := req.FromAddress != nil
+			if isFromAddress {
+				historyFilter = req.FromAddress
+			} else {
+				historyFilter = req.ToAddress
+			}
+
+			for _, addr := range historyFilter {
+
+				addrBytes := addr.Bytes()
+				blockNumbers, err := retrieveHistory(tx, addr, fromBlock, toBlock)
+				if err != nil {
+					return err
+				}
+
+				for _, num := range blockNumbers {
+
+					block := rawdb.ReadBlockByNumber(api.dbReader, num)
+					senders := rawdb.ReadSenders(api.dbReader, block.Hash(), num)
+					txs := block.Transactions()
+					for i, tx := range txs {
+						if uint64(len(filteredHashes)) == maxTracesCount {
+							if uint64(len(filteredHashes)) == api.maxTraces {
+								return fmt.Errorf("too many traces found")
+							}
+							return nil
+						}
+
+						var to *common.Address
+						if tx.To() == nil {
+							to = &common.Address{}
+						} else {
+							to = tx.To()
+						}
+
+						if isFromAddress {
+							if !isAddressInFilter(to, req.ToAddress) {
+								continue
+							}
+							if bytes.Equal(senders[i].Bytes(), addrBytes) {
+								filteredHashes = append(filteredHashes, tx.Hash())
+								traceTypes = append(traceTypes, false)
+							}
+						} else if bytes.Equal(to.Bytes(), addrBytes) {
```

Comment:
You can replace the comparison with `to == addr`

Reply:
In the next PR...